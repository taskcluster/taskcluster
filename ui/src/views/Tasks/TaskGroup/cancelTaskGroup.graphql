mutation CancelTaskGroup($taskGroupId: ID!) {
  cancelTaskGroup(taskGroupId: $taskGroupId) {
    taskGroupSize
    cancelledCount
    taskGroupId
  }
}
