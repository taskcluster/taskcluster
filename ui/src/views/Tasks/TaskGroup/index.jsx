import { hot } from 'react-hot-loader';
import React, { Component } from 'react';
import { graphql, withApollo } from 'react-apollo';
import dotProp from 'dot-prop-immutable';
import { isEmpty } from 'ramda';
import jsonSchemaDefaults from 'json-schema-defaults';
import { safeDump } from 'js-yaml';
import { withStyles } from '@material-ui/core/styles';
import classNames from 'classnames';
import Spinner from '@mozilla-frontend-infra/components/Spinner';
import HammerIcon from 'mdi-react/HammerIcon';
import SpeedDial from '../../../components/SpeedDial';
import SpeedDialAction from '../../../components/SpeedDialAction';
import Dashboard from '../../../components/Dashboard';
import Search from '../../../components/Search';
import DialogAction from '../../../components/DialogAction';
import HelpView from '../../../components/HelpView';
import TaskGroupProgress from '../../../components/TaskGroupProgress';
import TaskGroupTable from '../../../components/TaskGroupTable';
import TaskActionForm from '../../../components/TaskActionForm';
import {
  TASK_GROUP_PAGE_SIZE,
  TASK_GROUP_POLLING_INTERVAL,
  VALID_TASK,
  ACTIONS_JSON_KNOWN_KINDS,
  INITIAL_CURSOR,
} from '../../../utils/constants';
import db from '../../../utils/db';
import ErrorPanel from '../../../components/ErrorPanel';
import taskGroupQuery from './taskGroup.graphql';
import submitTaskAction from '../submitTaskAction';

const updateTaskGroupIdHistory = id => {
  if (!VALID_TASK.test(id)) {
    return;
  }

  db.taskGroupIdsHistory.put({ taskGroupId: id });
};

let previousCursor;

@hot(module)
@withApollo
@graphql(taskGroupQuery, {
  options: props => ({
    pollInterval: TASK_GROUP_POLLING_INTERVAL,
    errorPolicy: 'all',
    variables: {
      taskGroupId: props.match.params.taskGroupId,
      taskGroupConnection: {
        limit: TASK_GROUP_PAGE_SIZE,
      },
      taskActionsFilter: {
        kind: {
          $in: ACTIONS_JSON_KNOWN_KINDS,
        },
        context: {
          $or: [{ $size: 0 }, { $size: 1 }],
        },
      },
    },
  }),
})
@withStyles(theme => ({
  warningPanel: {
    ...theme.mixins.warningPanel,
  },
}))
export default class TaskGroup extends Component {
  static getDerivedStateFromProps(props, state) {
    const taskGroupId = props.match.params.taskGroupId || '';
    const { taskActions, taskGroup } = props.data;
    const groupActions = [];
    const actionInputs = state.actionInputs || {};
    const actionData = state.actionData || {};
    const taskGroupLoaded = taskGroup && !taskGroup.pageInfo.hasNextPage;
    // Make sure data is not from another task group which
    // can happen when a user searches for a different task group
    const isFromSameTaskGroupId =
      taskGroup && taskGroup.edges[0]
        ? taskGroup.edges[0].node.taskGroupId === taskGroupId
        : true;

    if (
      isFromSameTaskGroupId &&
      taskGroupId !== state.previousTaskGroupId &&
      taskActions
    ) {
      updateTaskGroupIdHistory(taskGroupId);
      taskActions.actions
        .filter(action => isEmpty(action.context))
        .forEach(action => {
          const schema = action.schema || {};

          // if an action with this name has already been selected,
          // don't consider this version
          if (!groupActions.some(({ name }) => name === action.name)) {
            groupActions.push(action);
            actionInputs[action.name] = safeDump(
              jsonSchemaDefaults(schema) || {}
            );
            actionData[action.name] = {
              action,
            };
          }
        });

      return {
        groupActions,
        actionInputs,
        actionData,
        previousTaskGroupId: taskGroupId,
        taskGroupLoaded,
      };
    }

    return {
      taskGroupLoaded: isFromSameTaskGroupId ? taskGroupLoaded : false,
    };
  }

  constructor(props) {
    super(props);

    previousCursor = INITIAL_CURSOR;
  }

  state = {
    filter: null,
    // eslint-disable-next-line react/no-unused-state
    previousTaskGroupId: '',
    groupActions: [],
    actionLoading: false,
    actionInputs: {},
    actionData: {},
    dialogOpen: false,
    selectedAction: null,
    dialogError: null,
    taskGroupLoaded: false,
  };

  componentDidUpdate(prevProps) {
    const {
      data: { taskGroup },
      match: {
        params: { taskGroupId },
      },
    } = this.props;

    if (prevProps.match.params.taskGroupId !== taskGroupId) {
      updateTaskGroupIdHistory(taskGroupId);
    }

    if (
      taskGroup &&
      previousCursor === taskGroup.pageInfo.cursor &&
      taskGroup.pageInfo.hasNextPage
    ) {
      this.fetchMoreTasks();
    }
  }

  handleActionClick = ({ currentTarget: { name } }) => {
    const { action } = this.state.actionData[name];

    this.setState({ dialogOpen: true, selectedAction: action });
  };

  handleActionComplete = taskId => {
    this.handleActionDialogClose();
    this.handleActionTaskComplete(taskId);
  };

  handleActionDialogClose = () => {
    this.setState({
      dialogOpen: false,
      selectedAction: null,
      dialogError: null,
      actionLoading: false,
    });
  };

  handleActionError = e => {
    this.setState({ dialogError: e, actionLoading: false });
  };

  handleActionSubmit = ({ name }) => async () => {
    this.preRunningAction();

    const { taskActions, task } = this.props.data;
    const { actionInputs, actionData } = this.state;
    const form = actionInputs[name];
    const { action } = actionData[name];
    const taskId = await submitTaskAction({
      task,
      taskActions,
      form,
      action,
      apolloClient: this.props.client,
    });

    return taskId;
  };

  handleActionTaskComplete = taskId => {
    this.props.history.push(`/tasks/${taskId}`);
  };

  handleFormChange = (value, name) =>
    this.setState({
      actionInputs: {
        // eslint-disable-next-line react/no-access-state-in-setstate
        ...this.state.actionInputs,
        [name]: value,
      },
    });

  handleStatusClick = async ({ currentTarget: { name } }) => {
    const filter = this.state.filter === name ? null : name;

    this.setState({ filter });
  };

  handleTaskGroupSearchSubmit = taskGroupId => {
    if (this.props.match.params.taskGroupId === taskGroupId) {
      return;
    }

    this.props.history.push(`/tasks/groups/${taskGroupId}`);
  };

  fetchMoreTasks = () => {
    const {
      data,
      match: {
        params: { taskGroupId },
      },
    } = this.props;
    const { fetchMore, taskGroup } = data;

    fetchMore({
      variables: {
        taskGroupId,
        taskGroupConnection: {
          limit: TASK_GROUP_PAGE_SIZE,
          cursor: taskGroup.pageInfo.nextCursor,
          previousCursor: taskGroup.pageInfo.cursor,
        },
        taskActionsFilter: {
          kind: {
            $in: ACTIONS_JSON_KNOWN_KINDS,
          },
          context: {
            $or: [{ $size: 0 }, { $size: 1 }],
          },
        },
      },
      updateQuery(previousResult, { fetchMoreResult, variables }) {
        if (variables.taskGroupConnection.previousCursor === previousCursor) {
          const { edges, pageInfo } = fetchMoreResult.taskGroup;

          if (!pageInfo.hasNextPage) {
            // Resetting to the initial cursor will allow us to
            // capture updates since the query has a polling interval
            previousCursor = INITIAL_CURSOR;
          } else {
            previousCursor = variables.taskGroupConnection.cursor;
          }

          if (!edges.length) {
            return previousResult;
          }

          const result = dotProp.set(previousResult, 'taskGroup', taskGroup =>
            dotProp.set(
              dotProp.set(
                taskGroup,
                'edges',
                previousResult.taskGroup.edges.concat(edges)
              ),
              'pageInfo',
              pageInfo
            )
          );

          return result;
        }
      },
    });
  };

  preRunningAction = () => {
    this.setState({ dialogError: null, actionLoading: true });
  };

  render() {
    const {
      groupActions,
      filter,
      actionLoading,
      dialogOpen,
      selectedAction,
      actionInputs,
      dialogError,
      taskGroupLoaded,
    } = this.state;
    const {
      description,
      match: {
        params: { taskGroupId },
      },
      data: { taskGroup, error, loading },
      classes,
    } = this.props;

    return (
      <Dashboard
        title="Task Group"
        helpView={<HelpView description={description} />}
        search={<Search onSubmit={this.handleTaskGroupSearchSubmit} />}>
        <ErrorPanel
          error={error}
          warning={Boolean(taskGroup)}
          className={classNames({
            [classes.warningPanel]: Boolean(taskGroup),
          })}
        />
        {taskGroup && (
          <TaskGroupProgress
            taskGroup={taskGroup}
            taskGroupId={taskGroupId}
            filter={filter}
            onStatusClick={this.handleStatusClick}
          />
        )}
        <br />
        {!error && !taskGroupLoaded && <Spinner loading />}
        {!loading && taskGroupLoaded && (
          <TaskGroupTable filter={filter} taskGroupConnection={taskGroup} />
        )}
        {!loading && groupActions && groupActions.length ? (
          <SpeedDial>
            {groupActions.map(action => (
              <SpeedDialAction
                requiresAuth
                tooltipOpen
                key={action.title}
                ButtonProps={{
                  name: action.name,
                  disabled: actionLoading,
                }}
                icon={<HammerIcon />}
                tooltipTitle={action.title}
                onClick={this.handleActionClick}
              />
            ))}
          </SpeedDial>
        ) : null}
        {dialogOpen && (
          <DialogAction
            fullScreen={Boolean(selectedAction.schema)}
            open={dialogOpen}
            error={dialogError}
            onSubmit={this.handleActionSubmit(selectedAction)}
            onComplete={this.handleActionComplete}
            onError={this.handleActionError}
            onClose={this.handleActionDialogClose}
            title={selectedAction.title}
            body={
              <TaskActionForm
                action={selectedAction}
                form={actionInputs[selectedAction.name]}
                onFormChange={this.handleFormChange}
              />
            }
            confirmText={selectedAction.title}
          />
        )}
      </Dashboard>
    );
  }
}
