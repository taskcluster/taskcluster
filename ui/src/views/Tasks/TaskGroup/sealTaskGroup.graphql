mutation SealTaskGroup($taskGroupId: ID!) {
  sealTaskGroup(taskGroupId: $taskGroupId) {
    taskGroupId
    schedulerId
    expires
    sealed
  }
}
