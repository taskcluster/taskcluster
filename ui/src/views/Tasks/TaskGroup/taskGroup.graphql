query TaskGroup($taskGroupId: ID!, $taskGroupConnection: PageConnection, $filter: JSON, $taskActionsFilter: JSON) {
  taskGroup(taskGroupId: $taskGroupId, connection: $taskGroupConnection, filter: $filter) {
    pageInfo {
      hasNextPage
      hasPreviousPage
      cursor
      previousCursor
      nextCursor
    }

    taskGroup {
      taskGroupId
      schedulerId
      expires
      sealed
    }

    edges {
      node {
        taskId
        metadata {
          name
        }
        taskGroupId
        status {
          state
          runs {
            runId
            started
            resolved
          }
        }

      }
    }
  }

  task(taskId: $taskGroupId) {
    taskQueueId
    schedulerId
    taskId
    taskGroupId
    dependencies
    requires
    routes
    priority
    retries
    created
    deadline
    scopes
    payload
    metadata {
      name
      description
      owner
      source
    }
    tags
    extra
  }

  taskActions(taskGroupId: $taskGroupId, filter: $taskActionsFilter) {
    actions
    variables
    version
  }
}
