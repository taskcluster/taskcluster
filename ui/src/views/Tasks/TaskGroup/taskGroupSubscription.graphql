subscription TaskGroupSubscription($taskGroupId: ID!, $subscriptions: [TaskSubscriptions]!) {
  tasksSubscriptions(taskGroupId: $taskGroupId, subscriptions: $subscriptions) {
    state
    runs {
      # Required to properly update the cache
      runId
      started
      resolved
    }
    # Required to properly update the cache
    taskId
    taskGroupId
    task {
      taskId
      metadata {
        name
      }
      taskGroupId
    }
  }
}
