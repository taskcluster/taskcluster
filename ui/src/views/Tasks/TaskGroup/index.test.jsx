import React from 'react';
import { render, waitFor, act } from '@testing-library/react';
import { ApolloProvider } from 'react-apollo';
import setupClient from 'apollo-client-mock';
import { MemoryRouter } from 'react-router-dom';
import TaskGroup from './index';

const typeDefs = `
  schema {
    query: TaskGroup
    subscription: TaskGroup
  }
  type Task {
    taskId: ID
  }
  type TaskActions {
    id: String
  }
  type TaskGroup {
    taskGroup: ID
    task: ID
    taskActions: TaskActions
  }
  enum TaskSubscriptions {
    tasksDefined
    tasksPending
    tasksRunning
    tasksCompleted
    tasksFailed
    tasksException
  }
  subscription TaskGroupSubscription {
    tasksSubscriptions: TaskSubscriptions
    state: String
    taskId: ID
    taskGroupId: ID
    task: String
  }
`;
const defaultMocks = {
  TaskGroup: () => ({
    taskGroup: () => ({
      taskId: 'taskId',
      taskGroupId: 'taskId',
    }),
    task: () => ({
      taskId: 'taskId',
      taskGroupId: 'taskId',
    }),
    taskActions: () => ({}),
  }),
  TaskGroupSubscription: () => ({
    tasksSubscriptions: () => ({
      taskId: 'taskId',
      taskGroupId: 'taskGroupId',
      state: 'state',
      task: {},
    }),
  }),
};

describe('TaskGroup page', () => {
  it('should render TaskGroup page', async () => {
    const createClient = setupClient(defaultMocks, typeDefs);
    const location = {
      hash: '#term',
    };

    await act(async () => {
      const { asFragment } = render(
        <MemoryRouter keyLength={0}>
          <ApolloProvider client={createClient()}>
            <TaskGroup
              match={{ params: { taskGroupId: 'aI8bvUB2SDmpHVqTUOFCWw' } }}
              location={location}
            />
          </ApolloProvider>
        </MemoryRouter>
      );

      await waitFor(() => {});
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
