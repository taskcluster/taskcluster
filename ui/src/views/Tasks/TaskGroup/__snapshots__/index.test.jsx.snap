// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskGroup page should render TaskGroup page 1`] = `<DocumentFragment />`;
